File: atak/ATAKMapEngine/lib/src/main/java/com/atakmap/map/layer/raster/tilematrix/TileProxy.java
package com.atakmap.map.layer.raster.tilematrix;

import android.graphics.Bitmap;

import com.atakmap.coremap.concurrent.NamedThreadFactory;
import com.atakmap.coremap.maps.coords.GeoPoint;
import com.atakmap.map.contentservices.CacheRequest;
import com.atakmap.map.contentservices.CacheRequestListener;
import com.atakmap.map.layer.feature.geometry.Envelope;
import com.atakmap.map.layer.raster.controls.TileCacheControl;
import com.atakmap.map.projection.Projection;
import com.atakmap.map.projection.ProjectionFactory;
import com.atakmap.math.MathUtils;
import com.atakmap.math.PointD;
import com.atakmap.util.ReferenceCount;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.Comparator;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public final class TileProxy implements TileClient {
    final static Comparator<TileFetchTask> requestPriorityComparator = new Comparator<TileFetchTask>() {
        @Override
        public int compare(TileFetchTask a, TileFetchTask b) {
            // items with highest priority go to end of LIFO queue
            return a.priority-b.priority;
        }
    };

    final ReferenceCount<TileClient> client;
    final ReferenceCount<TileContainer> cache;
    final ExecutorService clientExecutor;
    final TileCacheControl.OnTileUpdateListener[] listener;
    final Projection proj;
    long expiry;

    int priority = 0;

    ArrayList<TileFetchTask> downloadQueue = new ArrayList<>();

    PrioritizerImpl prioritizer = new PrioritizerImpl();

    public TileProxy(TileClient client, TileContainer cache) {
        this(client,
             cache,
             System.currentTimeMillis()-(24L*60L*60L*1000L), // 24 hours old
             Executors.newFixedThreadPool(3, new NamedThreadFactory("TileProxy[" + client.getName() + "]")));
    }

    TileProxy(TileClient client, TileContainer cache, long expiry, ExecutorService clientExecutor) {
        if(client == null)
            throw new IllegalArgumentException();
        if(cache == null)
            throw new IllegalArgumentException();
        if(clientExecutor == null)
            throw new IllegalArgumentException();

        this.client = new SharedTileMatrix<>(client);
        this.cache = new SharedTileMatrix<>(cache);
        this.expiry = expiry;
        this.clientExecutor = clientExecutor;
        this.listener = new TileCacheControl.OnTileUpdateListener[1];

        this.proj = ProjectionFactory.getProjection(client.getSRID());
    }

    /**
     * Requests that the client abort the download of the specified tile
     * @param zoom
     * @param x
     * @param y
     */
    public void abortTile(int zoom, int x, int y) {
        synchronized (downloadQueue) {
            // XXX - we could also just mark the tile as aborted, however, that
            //       would potentially lead to a situation where we're taking
            //       up a lot of slots in the circular buffer with tiles that
            //       will aren't downloadable
            final int idx = findQueuedRequestIndexNoSync(zoom, x, y);
            if(idx >= 0)
                downloadQueue.remove(idx);
        }
    }

    int findQueuedRequestIndexNoSync(int zoom, int x, int y) {
        for(int i = downloadQueue.size()-1; i >= 0; i--) {
            final TileFetchTask request = downloadQueue.get(i);
            if(request.x == x && request.y == y && request.z == zoom) {
                return i;
            }
        }
        return -1;
    }

    private void downloadTile(int zoom, int x, int y) {
        synchronized(downloadQueue) {
            final int idx = findQueuedRequestIndexNoSync(zoom, x, y);
            if(idx < 0) {
                downloadQueue.add(new TileFetchTask(++priority, client, cache, x, y, zoom, expiry, listener));
                clientExecutor.execute(new Downloader());
            }
        }
    }

    @Override
    public void clearAuthFailed() {
        client.value.clearAuthFailed();
    }

    @Override
    public void checkConnectivity() {
        client.value.checkConnectivity();
    }

    @Override
    public void cache(CacheRequest request, CacheRequestListener listener) {
        client.value.cache(request, listener);
    }

    @Override
    public int estimateTileCount(CacheRequest request) {
        return client.value.estimateTileCount(request);
    }

    @Override
    public <T> T getControl(Class<T> controlClazz) {
        if(TileCacheControl.class.isAssignableFrom(controlClazz))
            return (T)prioritizer;
        return client.value.getControl(controlClazz);
    }

    @Override
    public void getControls(Collection<Object> controls) {
        controls.add(prioritizer);
        client.value.getControls(controls);
    }

    @Override
    public String getName() {
        return client.value.getName();
    }

    @Override
    public int getSRID() {
        return client.value.getSRID();
    }

    @Override
    public ZoomLevel[] getZoomLevel() {
        return client.value.getZoomLevel();
    }

    @Override
    public double getOriginX() {
        return client.value.getOriginX();
    }

    @Override
    public double getOriginY() {
        return client.value.getOriginY();
    }

    @Override
    public Bitmap getTile(int zoom, int x, int y, Throwable[] error) {
        final Bitmap retval = cache.value.getTile(zoom, x, y, error);
        downloadTile(zoom, x, y);
        return retval;
    }

    @Override
    public byte[] getTileData(int zoom, int x, int y, Throwable[] error) {
        final byte[] retval = cache.value.getTileData(zoom, x, y, error);
        downloadTile(zoom, x, y);
        return retval;
    }

    @Override
    public Envelope getBounds() {
        return client.value.getBounds();
    }

    @Override
    public void dispose() {
        synchronized(listener) {
            listener[0] = null;
        }

        clientExecutor.shutdownNow();

        synchronized(downloadQueue) {
            for(int i = 0; i < downloadQueue.size(); i++) {
                downloadQueue.get(i).cache.dereference();
                downloadQueue.get(i).client.dereference();
            }
            downloadQueue.clear();
        }

        client.dereference();
        cache.dereference();
    }

    final static class SharedTileMatrix<T extends TileMatrix> extends ReferenceCount<T> {
        public SharedTileMatrix(T value) {
            super(value, true);
        }

        @Override
        protected void onDereferenced() {
            super.onDereferenced();
            value.dispose();
        }
    }

    final class Downloader implements Runnable {
        @Override
        public void run() {
            TileFetchTask task;
            synchronized (downloadQueue) {
                if(downloadQueue.isEmpty())
                    return;
                task = downloadQueue.remove(downloadQueue.size()-1);
            }
            task.run();
        }
    }

    final static class TileFetchTask implements Runnable {
        final ReferenceCount<TileClient> client;
        final ReferenceCount<TileContainer> cache;
        final int x;
        final int y;
        final int z;
        final long expiry;
        final TileCacheControl.OnTileUpdateListener[] callback;
        final Envelope bounds;
        final double centroidX;
        final double centroidY;
        final double centroidZ;
        final double radius;

        int priority;

        TileFetchTask(int priority, ReferenceCount<TileClient> client, ReferenceCount<TileContainer> cache, int x, int y, int z, long expiry, TileCacheControl.OnTileUpdateListener[] l) {
            this.priority = priority;
            this.client = client;
            this.cache = cache;
            this.x = x;
            this.y = y;
            this.z = z;
            this.expiry = expiry;
            this.callback = l;

            this.client.reference();
            this.cache.reference();

            bounds = TileMatrix.Util.getTileBounds(client.value, z, x, y);
            centroidX = (bounds.minX+bounds.maxX)/2d;
            centroidY = (bounds.minY+bounds.maxY)/2d;
            centroidZ = (bounds.minZ+bounds.maxZ)/2d;
            radius = MathUtils.distance(centroidX, centroidY, centroidZ, bounds.maxX, bounds.maxY, bounds.maxZ);
        }

        @Override
        public void run() {
            try {
                final long expiration = cache.value.getTileExpiration(z, x, y);
                // if the tile is considered expired, download
                if(expiration < expiry) {
                    byte[] data = client.value.getTileData(z, x, y, null);
                    if (data != null) {
                        cache.value.setTile(z, x, y, data, System.currentTimeMillis());
                        // signal update
                        synchronized (callback) {
                            if (callback[0] != null)
                                callback[0].onTileUpdated(z, x, y);
                        }
                    }
                }
            } catch(Throwable t) {
            } finally {
                client.dereference();
                cache.dereference();
            }
        }
    }

    final class PrioritizerImpl implements TileCacheControl, Comparator<TileFetchTask> {

        GeoPoint p;
        PointD xyz0 = new PointD(0d, 0d, 0d);
        PointD xyz1 = new PointD(0d, 0d, 0d);
        PointD xyz2 = new PointD(0d, 0d, 0d);

        @Override
        public int compare(TileFetchTask a, TileFetchTask b) {
            final double da = MathUtils.distance(a.centroidX, a.centroidY, a.centroidZ, xyz0.x, xyz0.y, xyz0.z);
            final double db = MathUtils.distance(b.centroidX, b.centroidY, b.centroidZ, xyz0.x, xyz0.y, xyz0.z);
            if(da <= a.radius && db <= b.radius)
                return b.z-a.z;
            else if(da <= a.radius)
                return 1;
            else if(db < b.radius)
                return -1;
            // both radii are greater
            else if((da-a.radius) < (db-b.radius))
                return 1;
            else if((da-a.radius) > (db-b.radius))
                return -1;
            else if(a.z < b.z)
                return 1;
            else if(a.z > b.z)
                return -1;
            else
                return a.priority-b.priority;
        }

        @Override
        public void prioritize(GeoPoint p) {
            this.p = (p != null) ? new GeoPoint(p) : null;
            if(this.p != null) {
                proj.forward(p, xyz0);
                proj.forward(p, xyz1);
                proj.forward(p, xyz2);
            }

            if(this.p != null)
            synchronized(downloadQueue) {
                Collections.sort(downloadQueue, this);
            }
        }

        @Override
        public void abort(int level, int x, int y) {
            abortTile(level, x, y);
        }

        @Override
        public boolean isQueued(int level, int x, int y) {
            synchronized(downloadQueue) {
                return findQueuedRequestIndexNoSync(level, x, y) >= 0;
            }
        }

        @Override
        public void setOnTileUpdateListener(OnTileUpdateListener l) {
            synchronized(listener) {
                listener[0] = l;
            }
        }

        @Override
        public void expireTiles(long expiry) {
            TileProxy.this.expiry = expiry;
        }
    }
}

File: takkernel/engine/src/main/cpp/raster/tilematrix/TileProxy.cpp
#ifdef MSVC
#include "raster/tilematrix/TileProxy.h"

#include <algorithm>

#include "core/ProjectionFactory3.h"
#include "math/Vector4.h"
#include "renderer/raster/TileClientControl.h"
#include "thread/Lock.h"

using namespace TAK::Engine::Raster::TileMatrix;

using namespace TAK::Engine::Core;
using namespace TAK::Engine::Math;
using namespace TAK::Engine::Port;
using namespace TAK::Engine::Renderer::Raster;
using namespace TAK::Engine::Thread;
using namespace TAK::Engine::Util;

namespace
{
    int64_t defaultExpiry() NOTHROWS
    {
        return Platform_systime_millis() - (24LL * 60LL * 60LL * 1000LL);
    }
}

class TileProxy::PrioritizerImpl : public TileCacheControl
{
public :
    PrioritizerImpl(TileProxy &owner) NOTHROWS;
    ~PrioritizerImpl() NOTHROWS;
public :
    void prioritize(const TAK::Engine::Core::GeoPoint2 &p) NOTHROWS override;
    void abort(const std::size_t level, const std::size_t x, const std::size_t y) NOTHROWS override;
    bool isQueued(const std::size_t level, const std::size_t x, const std::size_t y) NOTHROWS override;
    void setOnTileUpdateListener(OnTileUpdateListener *l) NOTHROWS override;
    void expireTiles(const int64_t expiry) NOTHROWS override;
private :
    TileProxy& owner;
};
class TileProxy::ClientControlImpl : public TileClientControl
{
public :
    ClientControlImpl(TileProxy &owner) NOTHROWS;
public :
    void setOfflineOnlyMode(bool offlineOnly) override;
    bool isOfflineOnlyMode() override;
    void refreshCache() override;
    void setCacheAutoRefreshInterval(int64_t milliseconds) override;
    int64_t getCacheAutoRefreshInterval() override;
private :
    TileProxy& owner;
    bool offlineOnly;
    int64_t refreshInterval;
};

TileProxy::TileProxy(TileClientPtr &&client, TileContainerPtr &&cache) NOTHROWS :
    TileProxy(std::shared_ptr<TileClient>(std::move(client)), std::shared_ptr<TileContainer>(std::move(cache)), defaultExpiry())
{}
TileProxy::TileProxy(const std::shared_ptr<TileClient>& client, const std::shared_ptr<TileContainer>& cache) NOTHROWS :
    TileProxy(client, cache, defaultExpiry())
{}
TileProxy::TileProxy(const std::shared_ptr<TileClient>& client_, const std::shared_ptr<TileContainer>& cache_, const int64_t expiry_) NOTHROWS :
    proj(nullptr, nullptr),
    expiry(expiry_),
    downloadQueue(std::make_shared<std::vector<TileFetchTask>>()),
    monitor(std::make_shared<Monitor>()),
    threadPool(nullptr, nullptr),
    detached(std::make_shared<bool>(false)),
    callback(std::make_shared<TileUpdateListener>())
{
    impl.client = client_;
    impl.cache = cache_;
    ProjectionFactory3_create(proj, impl.client->getSRID());

    prioritizer = std::make_shared<PrioritizerImpl>(*this);
    clientControl = std::make_unique<ClientControlImpl>(*this);
}
TileProxy::~TileProxy() NOTHROWS
{
    Monitor::Lock lock(*monitor);
    // Detach all threads in the pool. This allows for immediate destruct,
    // client and cache will only be kept open if 1) referenced externally or
    // 2) ongoing active download/cache operation
    if (threadPool) {
        threadPool->detachAll();
        threadPool.reset();
    }

    // mark as detached
    *detached = true;

    // clear the queue
    downloadQueue->clear();
    prioritizer.reset();

    // notify
    lock.broadcast();
}

TAKErr TileProxy::abortTile(const std::size_t zoom, const std::size_t x, const std::size_t y) NOTHROWS
{
    {
        Monitor::Lock lock(*monitor);
        TE_CHECKRETURN_CODE(lock.status);

        // XXX - we could also just mark the tile as aborted, however, that
        //       would potentially lead to a situation where we're taking
        //       up a lot of slots in the circular buffer with tiles that
        //       will aren't downloadable
        std::size_t idx;
        if (findQueuedRequestIndexNoSync(&idx, zoom, x, y) == TE_Ok)
            downloadQueue->erase(downloadQueue->begin() + idx);
    }
    return TE_Ok;
}
TAKErr TileProxy::clearAuthFailed() NOTHROWS
{
    return impl.client->clearAuthFailed();
}
TAKErr TileProxy::checkConnectivity() NOTHROWS
{
    return impl.client->checkConnectivity();
}
TAKErr TileProxy::cache(const CacheRequest &request, std::shared_ptr<CacheRequestListener> &listener) NOTHROWS
{
    return impl.client->cache(request, listener);
}
TAKErr TileProxy::estimateTilecount(int *count, const CacheRequest &request) NOTHROWS
{
    return impl.client->estimateTilecount(count, request);
}
const char* TileProxy::getName() const NOTHROWS
{
    return impl.client->getName();
}
int TileProxy::getSRID() const NOTHROWS
{
    return impl.client->getSRID();
}
TAKErr TileProxy::getZoomLevel(Port::Collection<ZoomLevel>& value) const NOTHROWS
{
    return impl.client->getZoomLevel(value);
}
double TileProxy::getOriginX() const NOTHROWS
{
    return impl.client->getOriginX();
}
double TileProxy::getOriginY() const NOTHROWS
{
    return impl.client->getOriginY();
}
TAKErr TileProxy::getTile(Renderer::BitmapPtr& result, const std::size_t zoom, const std::size_t x, const std::size_t y) NOTHROWS
{
    const TAKErr code = impl.cache->getTile(result, zoom, x, y);
    downloadTile(zoom, x, y);
    return code;
}
TAKErr TileProxy::getTileData(std::unique_ptr<const uint8_t, void (*)(const uint8_t*)>& value, std::size_t* len,
        const std::size_t zoom, const std::size_t x, const std::size_t y) NOTHROWS
{
    const TAKErr code = impl.cache->getTileData(value, len, zoom, x, y);
    downloadTile(zoom, x, y);
    return code;
}
TAKErr TileProxy::getBounds(Feature::Envelope2* value) const NOTHROWS
{
    return impl.client->getBounds(value);
}
TAKErr TileProxy::getControl(Control* control, const char* controlName) NOTHROWS
{
    if (!control)
        return TE_InvalidArg;

    if(String_strcmp(TileCacheControl_getType(), controlName) == 0) {
        control->value = static_cast<TileCacheControl*>(prioritizer.get());
        control->type = TileCacheControl_getType();
        return TE_Ok;
    } else if(String_strcmp(TileClientControl_getType(), controlName) == 0) {
        control->value = static_cast<TileClientControl*>(clientControl.get());
        control->type = TileClientControl_getType();
        return TE_Ok;
    }
    return TE_Done;
}
TAKErr TileProxy::getControls(Collection<Control>& controls) NOTHROWS
{
    TAKErr code(TE_Ok);
    code = controls.add(Control{TileCacheControl_getType(), static_cast<TileCacheControl*>(prioritizer.get())});
    code = controls.add(Control{TileClientControl_getType(), static_cast<TileClientControl*>(clientControl.get())});
    TE_CHECKRETURN_CODE(code);

    return code;
}
TAKErr TileProxy::findQueuedRequestIndexNoSync(std::size_t* value, const std::size_t zoom, const std::size_t x, const std::size_t y) const NOTHROWS
{
    if (!downloadQueue->empty()) {
        for (std::size_t i = downloadQueue->size(); i > 0; i--) {
            const std::size_t idx = i - 1u;
            const auto &request = (*downloadQueue)[idx];
            if (request.tileIndex.x == x && request.tileIndex.y == y && request.tileIndex.z == zoom) {
                *value = idx;
                return TE_Ok;
            }
        }
    }
    return TE_Done;
}
TAKErr TileProxy::downloadTile(const std::size_t zoom, const std::size_t x, const std::size_t y) NOTHROWS
{
    // nothing to do if offline mode
    if (clientControl->isOfflineOnlyMode())
        return TE_Done;

    {
        Monitor::Lock lock(*monitor);
        TE_CHECKRETURN_CODE(lock.status);
        std::size_t idx;
        if(findQueuedRequestIndexNoSync(&idx, zoom, x, y) == TE_Done) {
            if(!threadPool) {
                const std::size_t numThreads = 3u;
                std::vector<void *> workerData;
                workerData.reserve(numThreads);
                for(std::size_t i = 0u; i < numThreads; i++) {
                    std::unique_ptr<TileFetchWorker> worker(new TileFetchWorker());
                    worker->downloadQueue = downloadQueue;
                    worker->monitor = monitor;
                    worker->detached = detached;
                    worker->callback = callback;
                    workerData.push_back(worker.release());
                }
                if (ThreadPool_create(threadPool, numThreads, downloadThread, &workerData.at(0)) != TE_Ok)
                    return TE_Err;
            }
            TileFetchTask task;
            task.priority = ++priority;
            task.source = impl.client;
            task.sink = impl.cache;
            task.tileIndex = TAK::Engine::Math::Point2<std::size_t>(x, y, zoom);
            task.expiry = expiry;
            TileMatrix_getTileBounds(&task.bounds, *impl.client, (int)zoom, (int)x, (int)y);
            task.centroid.x = (task.bounds.minX+task.bounds.maxX)/2.0;
            task.centroid.y = (task.bounds.minY+task.bounds.maxY)/2.0;
            task.centroid.z = 0.0;
            task.radius = Vector2_length(Point2<double>(task.centroid.x-task.bounds.maxX, task.centroid.y-task.bounds.maxY, 0.0));

            downloadQueue->push_back(task);
            lock.signal();
        }
    }
    return TE_Ok;
}
void *TileProxy::downloadThread(void *opaque) NOTHROWS
{
    std::unique_ptr<TileFetchWorker> worker(static_cast<TileFetchWorker*>(opaque));
    while(true) {
        TileFetchTask task;
        {
            Monitor::Lock lock(*worker->monitor);
            if (*worker->detached)
                break;
            if(worker->downloadQueue->empty()) {
                lock.wait();
                continue;
            }

            task = worker->downloadQueue->back();
            worker->downloadQueue->pop_back();
        }

        // execute task
        int64_t expiration(-1LL);
        {
            const std::shared_ptr<TileContainer> cache = task.sink.lock();
            expiration = cache->getTileExpiration(task.tileIndex.z, task.tileIndex.x, task.tileIndex.y);
        }

        if (expiration >= task.expiry)
            continue;

        // the tile is considered expired, download
        const std::shared_ptr<TileClient> client = task.source.lock();
        if (!client)
            continue;

        std::unique_ptr<const uint8_t, void(*)(const uint8_t*)> data(nullptr, nullptr);
        std::size_t len;
        if (client->getTileData(data, &len, task.tileIndex.z, task.tileIndex.x, task.tileIndex.y) != TE_Ok)
            continue;

        {
            const std::shared_ptr<TileContainer> cache = task.sink.lock();
            if (!cache || cache->setTile(task.tileIndex.z, task.tileIndex.x, task.tileIndex.y, data.get(), len, Platform_systime_millis()) != TE_Ok)
                continue;
            // signal update
            {
                Lock lock(worker->callback->mutex);
                if (worker->callback->value)
                    worker->callback->value->onTileUpdated(task.tileIndex.z, task.tileIndex.x, task.tileIndex.y);
            }
        }
    }

    return nullptr;
}

TileProxy::PrioritizerImpl::PrioritizerImpl(TileProxy &owner_) NOTHROWS :
    owner(owner_)
{
}
TileProxy::PrioritizerImpl::~PrioritizerImpl() NOTHROWS
{
    Lock lock(owner.callback->mutex);
    owner.callback->value = nullptr;
}

void TileProxy::PrioritizerImpl::prioritize(const TAK::Engine::Core::GeoPoint2 &lla) NOTHROWS
{
    Point2<double> xyz;
    if (owner.proj->forward(&xyz, lla) != TE_Ok)
        return;

    Monitor::Lock lock(*owner.monitor);
    std::sort(owner.downloadQueue->begin(), owner.downloadQueue->end(), [xyz](const TileFetchTask& a, const TileFetchTask& b)
    {
        do {
            // XXX - use distance-squared for efficiency
            const double da = Vector2_length(Vector2_subtract(a.centroid, xyz));
            const double db = Vector2_length(Vector2_subtract(b.centroid, xyz));
            if (da <= a.radius && db <= b.radius)
                break; // priority focus is contained within both bounding circles, default priority
            else if (da <= a.radius)
                return false; // contained in A bounding circle
            else if (db <= b.radius)
                return true; // contained in B bounding circle
            // both radii are greater
            else if ((da - a.radius) < (db - b.radius))
                return false; // closer to A
            else if ((da - a.radius) > (db - b.radius))
                return true; // closer to B
        } while (false);

        // default prioritization, based first on lo->hi res, then insert order
        if (a.tileIndex.z < b.tileIndex.z)
            return false;
        else if (a.tileIndex.z > b.tileIndex.z)
            return true;
        else
            return (a.priority < b.priority);
    });
}
void TileProxy::PrioritizerImpl::abort(const std::size_t level, const std::size_t x, const std::size_t y) NOTHROWS
{
    owner.abortTile(level, x, y);
}
bool TileProxy::PrioritizerImpl::isQueued(const std::size_t level, const std::size_t x, const std::size_t y) NOTHROWS
{
    Monitor::Lock lock(*owner.monitor);
    std::size_t idx;
    return owner.findQueuedRequestIndexNoSync(&idx, level, x, y) != TE_Done;
}
void TileProxy::PrioritizerImpl::setOnTileUpdateListener(OnTileUpdateListener *l) NOTHROWS
{
    Lock lock(owner.callback->mutex);
    owner.callback->value = l;
}
void TileProxy::PrioritizerImpl::expireTiles(const int64_t expiry) NOTHROWS
{
    owner.expiry = expiry;
}

TileProxy::ClientControlImpl::ClientControlImpl(TileProxy &owner_) NOTHROWS :
    owner(owner_),
    offlineOnly(false),
    refreshInterval(0LL)
{}
void TileProxy::ClientControlImpl::setOfflineOnlyMode(bool value)
{
    offlineOnly = value;
}
bool TileProxy::ClientControlImpl::isOfflineOnlyMode()
{
    return offlineOnly;
}
void TileProxy::ClientControlImpl::refreshCache()
{
    owner.prioritizer->expireTiles(Platform_systime_millis()-1LL);
}
void TileProxy::ClientControlImpl::setCacheAutoRefreshInterval(int64_t milliseconds)
{
    refreshInterval = milliseconds;
}
int64_t TileProxy::ClientControlImpl::getCacheAutoRefreshInterval()
{
    return refreshInterval;
}
#endif